metricbeat/mb/lightmodules.go

```diff
@@ -0,0 +1,288 @@
+		}
+
+		return originalFactory(base)
+	}
+
+	return registration, nil
+}
+
+// baseModule does the configuration overrides in the base module configuration
+// taking into account the light metric set default configurations
+func (m *LightMetricSet) baseModule(from Module) (*BaseModule, error) {
+	baseModule := BaseModule{
+		name:   m.Module,
+		config: from.Config(),
+	}
+	var err error
+	// Set defaults
+	if baseModule.rawConfig, err = common.NewConfigFrom(m.Input.Defaults); err != nil {
+		return nil, errors.Wrap(err, "invalid input defaults")
+	}
+	// Copy values from user configuration
+	if err = from.UnpackConfig(baseModule.rawConfig); err != nil {
+		return nil, errors.Wrap(err, "failed to copy values from user configuration")
+	}
+	// Update module configuration
+	if err = baseModule.UnpackConfig(&baseModule.config); err != nil {
+		return nil, errors.Wrap(err, "failed to set module configuration")
+	}
+	return &baseModule, nil
+}
+
+func (r *LightModulesRegistry) readModule(moduleName string) (*LightModule, bool, error) {
+	modulePath, found := r.findModulePath(moduleName)
+	if !found {
+		return nil, false, nil
+	}
+
+	moduleConfig, err := r.readModuleConfig(modulePath)
+	if err != nil {
+		return nil, true, errors.Wrapf(err, "failed to load light module '%s' definition", moduleName)
+	}
+
+	metricSets := make(map[string]LightMetricSet)
+	for _, metricSet := range moduleConfig.MetricSets {
+		manifestPath := filepath.Join(filepath.Dir(modulePath), metricSet, manifestYML)
+
+		metricSetConfig, err := r.readMetricSetConfig(manifestPath)
+		if err != nil {
+			return nil, true, errors.Wrapf(err, "failed to load light metricset '%s/%s' definition", moduleName, metricSet)
+		}
+		metricSetConfig.Name = metricSet
+		metricSetConfig.Module = moduleName
+
+		metricSets[metricSet] = metricSetConfig
+	}
+
+	return &LightModule{Name: moduleName, MetricSets: metricSets}, true, nil
+}
+
+func (r *LightModulesRegistry) findModulePath(moduleName string) (string, bool) {
+	for _, dir := range r.paths {
+		p := filepath.Join(dir, moduleName, moduleYML)
+		if _, err := os.Stat(p); err == nil {
+			return p, true
+		}
+	}
+	return "", false
+}
+
+func (r *LightModulesRegistry) readModuleConfig(modulePath string) (*lightModuleConfig, error) {
+	c, err := common.LoadFile(modulePath)
+	if err != nil {
+		return nil, err
+	}
+
+	var moduleConfig lightModuleConfig
+	if err = c.Unpack(&moduleConfig); err != nil {
+		return nil, errors.Wrapf(err, "failed to parse light module definition from '%s'", modulePath)
+	}
+	return &moduleConfig, nil
+}
+
+func (r *LightModulesRegistry) readMetricSetConfig(manifestPath string) (ms LightMetricSet, err error) {
+	c, err := common.LoadFile(manifestPath)
+	if err != nil {
+		return ms, err
+	}
+
+	if err := c.Unpack(&ms); err != nil {
+		return ms, errors.Wrapf(err, "failed to parse metricset manifest from '%s'", manifestPath)
+	}
+	return
+}
+
+func (r *LightModulesRegistry) listModules() ([]string, error) {
```

Comment:
Maybe `moduleNames` (if they don't contain a path), `modulesPaths` (but there's a `path` package in Go so it might be misleading) or `modulesDirPaths`... Also, listModules is very imperative and it's asking an argument.

Think that when we read this call from somewhere else in the future the variable name should be very close to the function name. I mean if I call this from somewhere else it's gonna look like `moduleList, err := r.listModules()` but those `modulesList` is a list of modules names? paths? structs? I'd say structs while that `list` word is still confusing me.